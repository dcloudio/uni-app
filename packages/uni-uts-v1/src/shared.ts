import os from 'os'
import path from 'path'
import { parse } from 'jsonc-parser'

export function parseJson(jsonStr: string) {
  return parse(jsonStr)
}

export function once<T extends (...args: any[]) => any>(
  fn: T,
  ctx: unknown = null
): T {
  let res: any
  return ((...args: any[]) => {
    if (fn) {
      res = fn.apply(ctx, args)
      fn = null as any
    }
    return res
  }) as T
}

export const runByHBuilderX = once(() => {
  return (
    !!process.env.UNI_HBUILDERX_PLUGINS &&
    (!!process.env.RUN_BY_HBUILDERX || !!process.env.HX_Version)
  )
})

export const isInHBuilderX = once(() => {
  try {
    // eslint-disable-next-line no-restricted-globals
    const { name } = require(path.resolve(
      process.cwd(),
      '../about/package.json'
    ))
    if (name === 'about') {
      process.env.UNI_HBUILDERX_PLUGINS = path.resolve(process.cwd(), '..')
      return true
    }
  } catch (e) {
    // console.error(e)
  }
  return false
})

export function resolveSourceMapPath(
  outputDir?: string,
  platform?: UniApp.PLATFORM
) {
  let dir = platform || process.env.UNI_SUB_PLATFORM || process.env.UNI_PLATFORM
  if (dir === 'app-plus') {
    dir = 'app'
  }
  return path.resolve(
    outputDir || process.env.UNI_OUTPUT_DIR,
    '../.sourcemap/' + dir
  )
}

export const isWindows = os.platform() === 'win32'
export function normalizePath(id: string): string {
  return isWindows ? id.replace(/\\/g, '/') : id
}

function supportAutoInstallPlugin() {
  return !!process.env.HX_Version
}

export function installHBuilderXPlugin(plugin: string) {
  if (!supportAutoInstallPlugin()) {
    return
  }
  return console.error(
    `%HXRunUniAPPPluginName%${plugin}%HXRunUniAPPPluginName%`
  )
}
