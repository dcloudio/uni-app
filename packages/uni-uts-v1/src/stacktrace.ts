import { originalPositionFor } from './sourceMap'
import { relative } from './utils'

const splitRE = /\r?\n/
const uniModulesSwiftUTSRe =
  /(.*)index.swift:([0-9]+):([0-9]+):\s+error:\s+(.*)/

interface ParseUTSPluginStacktraceOptions {
  stacktrace: string
  sourceRoot: string
  sourceMapFile: string
}

export async function parseUTSSwiftPluginStacktrace({
  stacktrace,
  sourceRoot,
  sourceMapFile,
}: ParseUTSPluginStacktraceOptions) {
  const res: string[] = []
  const lines = stacktrace.split(splitRE)
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]
    const codes = await parseUTSStacktraceLine(
      line,
      uniModulesSwiftUTSRe,
      sourceMapFile,
      sourceRoot
    )
    if (codes && codes.length) {
      res.push(...codes)
    } else {
      res.push(line)
    }
  }
  return res.join('\n')
}

async function parseUTSStacktraceLine(
  lineStr: string,
  re: RegExp,
  sourceMapFile: string,
  sourceRoot: string
) {
  const uniModulesMatches = lineStr.match(re)
  if (!uniModulesMatches) {
    return
  }
  const lines: string[] = []
  const [, , line, column, message] = uniModulesMatches
  const originalPosition = await originalPositionFor({
    sourceMapFile,
    line: parseInt(line),
    column: parseInt(column),
    withSourceContent: true,
  })

  if (originalPosition.source && originalPosition.sourceContent) {
    lines.push(`${message}`)
    lines.push(
      `at ${originalPosition.source.split('?')[0]}:${originalPosition.line}:${
        originalPosition.column
      }`
    )
    if (originalPosition.line !== null && originalPosition.column !== null) {
      lines.push(
        generateCodeFrame(originalPosition.sourceContent, {
          line: originalPosition.line,
          column: originalPosition.column,
        }).replace(/\t/g, ' ')
      )
    }
  } else {
    lines.push(lineStr)
  }
  return lines
}

const range: number = 2

function posToNumber(
  source: string,
  pos: number | { line: number; column: number }
): number {
  if (typeof pos === 'number') return pos
  const lines = source.split(splitRE)
  const { line, column } = pos
  let start = 0
  for (let i = 0; i < line - 1; i++) {
    start += lines[i].length + 1
  }
  return start + column
}

export function generateCodeFrame(
  source: string,
  start: number | { line: number; column: number } = 0,
  end?: number
): string {
  start = posToNumber(source, start)
  end = end || start
  const lines = source.split(splitRE)
  let count = 0
  const res: string[] = []
  for (let i = 0; i < lines.length; i++) {
    count += lines[i].length + 1
    if (count >= start) {
      for (let j = i - range; j <= i + range || end > count; j++) {
        if (j < 0 || j >= lines.length) continue
        const line = j + 1
        res.push(
          `${line}${' '.repeat(Math.max(3 - String(line).length, 0))}|  ${
            lines[j]
          }`
        )
        const lineLength = lines[j].length
        if (j === i) {
          // push underline
          const pad = start - (count - lineLength) + 1
          const length = Math.max(
            1,
            end > count ? lineLength - pad : end - start
          )
          res.push(`   |  ` + ' '.repeat(pad) + '^'.repeat(length))
        } else if (j > i) {
          if (end > count) {
            const length = Math.max(Math.min(end - count, lineLength), 1)
            res.push(`   |  ` + '^'.repeat(length))
          }
          count += lineLength + 1
        }
      }
      break
    }
  }
  return res.join('\n')
}

export function parseUTSSyntaxError(error: any, inputDir: string): string {
  if (error instanceof Error) {
    error = error.message + (error.stack ? `\n` + error.stack : ``)
  }
  let msg = String(error).replace(/\t/g, ' ')
  let res: RegExpExecArray | null = null
  const syntaxErrorRe = /(,-\[(.*):(\d+):(\d+)\])/g
  while ((res = syntaxErrorRe.exec(msg))) {
    const [row, filename, line, column] = res.slice(1)
    msg = msg.replace(
      row,
      `at ${relative(filename.split('?')[0], inputDir)}:${
        parseInt(line) + 1
      }:${column}`
    )
  }
  return msg
}
