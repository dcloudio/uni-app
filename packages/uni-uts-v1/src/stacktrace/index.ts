import { relative } from '../utils'
import {
  type GenerateAppIOSJavaScriptRuntimeCodeFrameOptions,
  type GenerateJavaScriptRuntimeCodeFrameOptions,
  parseUTSJavaScriptRuntimeStacktrace,
} from './js'
import {
  type GenerateAppAndroidKotlinRuntimeCodeFrameOptions,
  type GenerateKotlinRuntimeCodeFrameOptions,
  parseUTSKotlinRuntimeStacktrace,
} from './kotlin'
import {
  type GenerateMiniProgramRuntimeCodeFrameOptions,
  MP_PLATFORMS,
  parseMiniProgramRuntimeStacktrace,
} from './mp'

export { parseUTSSwiftPluginStacktrace } from './swift'

export {
  parseUTSKotlinStacktrace,
  parseUTSKotlinRuntimeStacktrace,
} from './kotlin'

export { parseUTSJavaScriptRuntimeStacktrace } from './js'

export async function parseRuntimeStacktrace(
  stacktrace: string,
  options:
    | GenerateAppAndroidKotlinRuntimeCodeFrameOptions
    | GenerateAppIOSJavaScriptRuntimeCodeFrameOptions
    | GenerateMiniProgramRuntimeCodeFrameOptions
) {
  if (
    (options.platform === 'app-android' && options.language === 'kotlin') ||
    (options.platform === 'app-ios' && options.language === 'javascript')
  ) {
    return parseUTSRuntimeStacktrace(stacktrace, options)
  }
  // mp-weixin:   sourceMap可以合并映射（所以开发工具可以显示源码） 可以读取到sourceMap（下载js文件，解析里边的base64格式sourceMap）
  // mp-baidu:    sourceMap无法合并映射（所以开发工具无法显示源码） 可以读取到sourceMap 二次解析映射（也可以合并sourceMap吧）
  // mp-toutiao:  sourceMap可以合并映射（所以开发工具可以显示源码） 可以读取到sourceMap（下载js文件，解析里边的base64格式sourceMap）
  // mp-alipay:   sourceMap可以合并映射（所以开发工具可以显示源码） 可以读取到sourceMap（下载js文件，解析url格式的sourceMap，再根据url下载sourceMap）
  if (MP_PLATFORMS[options.platform]) {
    return parseMiniProgramRuntimeStacktrace(stacktrace, options)
  }
  // 其他小程序平台暂不处理，因为没法拿到小程序的sourceMap做合并映射
  return stacktrace
}

export function parseUTSRuntimeStacktrace(
  stacktrace: string,
  options:
    | GenerateKotlinRuntimeCodeFrameOptions
    | GenerateJavaScriptRuntimeCodeFrameOptions
) {
  if (options.language === 'kotlin') {
    return parseUTSKotlinRuntimeStacktrace(stacktrace, options)
  } else if (options.language === 'javascript') {
    return parseUTSJavaScriptRuntimeStacktrace(stacktrace, options)
  }
}

export function parseUTSSyntaxError(error: any, inputDir: string): string {
  if (error instanceof Error) {
    error = error.message
  }
  let msg = String(error).replace(/\t/g, ' ')
  let res: RegExpExecArray | null = null
  const syntaxErrorRe = /(,-\[(.*):(\d+):(\d+)\])/g
  while ((res = syntaxErrorRe.exec(msg))) {
    const [row, filename, line, column] = res.slice(1)
    msg = msg.replace(
      row,
      `at ${relative(filename.split('?')[0], inputDir)}:${
        parseInt(line) + 3
      }:${column}`
    )
  }
  return msg
}
