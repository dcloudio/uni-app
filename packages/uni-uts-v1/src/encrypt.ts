import path, { join, relative } from 'path'
import fs from 'fs-extra'
import type { APP_PLATFORM } from './manifest/utils'
import { normalizePath, resolveSourceMapPath } from './shared'
import {
  type KotlinCompilerServer,
  compileAndroidDex,
  createStderrListener,
  getUniModulesCacheJars,
  resolveDexFile,
  resolveJarPath,
} from './kotlin'
import {
  addPluginInjectApis,
  addPluginInjectComponents,
  getCompilerServer,
  requireUniHelpers,
} from './utils'
import { restoreDex } from './manifest'
import { sync } from 'fast-glob'
import { resolveDexCacheFile } from './manifest/dex'
import type { CompileResult } from './index'
import { hbuilderFormatter } from './stacktrace/kotlin'

export function isEncrypt(pluginDir: string) {
  return fs.existsSync(path.resolve(pluginDir, 'encrypt'))
}

export function isUTSModules(pluginDir: string) {
  return fs.existsSync(path.resolve(pluginDir, 'utssdk'))
}

function createRollupCommonjsCode(
  _pluginDir: string,
  pluginRelativeDir: string
) {
  // const name = makeLegalIdentifier(pluginRelativeDir)
  return `
export default uni.requireUTSPlugin('${normalizePath(pluginRelativeDir)}')
`
}
function createWebpackCommonjsCode(pluginRelativeDir: string) {
  return `
module.exports = uni.requireUTSPlugin('${normalizePath(pluginRelativeDir)}')
`
}

export async function compileEncrypt(
  pluginDir: string,
  isX = false
): Promise<CompileResult> {
  if (isX && !fs.existsSync(path.resolve(pluginDir, 'utssdk'))) {
    return compileEncryptByUniHelpers(pluginDir)
  }

  const inputDir = process.env.UNI_INPUT_DIR
  const outputDir = process.env.UNI_OUTPUT_DIR
  const utsPlatform = process.env.UNI_UTS_PLATFORM as APP_PLATFORM
  const isRollup = !!process.env.UNI_UTS_USING_ROLLUP
  const pluginRelativeDir = relative(inputDir, pluginDir)
  const outputPluginDir = normalizePath(join(outputDir, pluginRelativeDir))
  const isNative = isX && utsPlatform === 'app-android'
  let code = isNative
    ? 'export default {}'
    : isRollup
    ? createRollupCommonjsCode(pluginDir, pluginRelativeDir)
    : createWebpackCommonjsCode(pluginRelativeDir)
  if (process.env.NODE_ENV !== 'development') {
    // 生成wgt，无需复制加密插件目录
    const needCopy = !(process.env.UNI_APP_PRODUCTION_TYPE === 'WGT')
    if (needCopy) {
      // 复制插件目录
      fs.copySync(pluginDir, join(outputDir, pluginRelativeDir))
    }
    return {
      dir: outputPluginDir,
      code,
      deps: [] as string[],
      encrypt: true,
      inject_apis: [],
      scoped_slots: [],
      meta: { commonjs: { isCommonJS: true } },
    }
  }
  const cacheDir = process.env.HX_DEPENDENCIES_DIR!
  if (!isNative) {
    // 读取缓存目录的 js code
    const indexJsPath = resolveJsCodeCacheFilename(
      utsPlatform,
      cacheDir,
      pluginRelativeDir
    )
    if (fs.existsSync(indexJsPath)) {
      code = fs.readFileSync(indexJsPath, 'utf-8') + code
    } else {
      console.error(
        `uts插件[${path.basename(pluginDir)}]不存在，请重新打包自定义基座`
      )
    }
  } else {
    const jarPath = resolveJarCacheFilename(cacheDir, pluginRelativeDir)
    if (!fs.existsSync(jarPath)) {
      console.error(
        `uts插件[${path.basename(pluginDir)}]不存在，请重新打包自定义基座`
      )
    }
  }
  return {
    dir: outputPluginDir,
    code,
    deps: [] as string[],
    encrypt: true,
    inject_apis: [],
    scoped_slots: [],
    meta: { commonjs: { isCommonJS: true } },
  }
}

export function resolveJsCodeCacheFilename(
  platform: APP_PLATFORM,
  cacheDir: string,
  pluginRelativeDir: string
) {
  return join(cacheDir, platform, 'uts', pluginRelativeDir, 'index.js')
}

function resolveJarCacheFilename(cacheDir: string, pluginRelativeDir: string) {
  return join(cacheDir, 'app-android', 'uts', pluginRelativeDir, 'index.jar')
}

async function compileEncryptByUniHelpers(pluginDir: string) {
  // 目前仅Android会进来
  const cacheDir = process.env.HX_DEPENDENCIES_DIR || ''
  const inputDir = process.env.UNI_INPUT_DIR
  const outputDir = process.env.UNI_OUTPUT_DIR
  const pluginRelativeDir = relative(inputDir, pluginDir)
  const outputPluginDir = normalizePath(join(outputDir, pluginRelativeDir))
  // 读取缓存文件
  const cachePluginDir = path.resolve(
    process.env.UNI_MODULES_ENCRYPT_CACHE_DIR!,
    pluginRelativeDir
  )

  const cacheFile = resolveDexCacheFile(pluginRelativeDir, cacheDir)
  if (process.env.NODE_ENV === 'development' && cacheFile) {
    // 已有缓存
    restoreDex(pluginRelativeDir, cacheDir, outputDir, true)
    const assets = path.resolve(cachePluginDir, 'assets')
    if (fs.existsSync(assets)) {
      fs.copySync(assets, path.resolve(outputDir, pluginRelativeDir, 'assets'))
    }
    return {
      dir: outputPluginDir,
      code: 'export default {}',
      deps: [] as string[],
      encrypt: true,
      inject_apis: [],
      scoped_slots: [],
    }
  }

  const pkg = require(path.resolve(cachePluginDir, 'package.json'))
  if (process.env.NODE_ENV !== 'development') {
    // 生成wgt，无需复制加密插件目录
    const needCopy = !(process.env.UNI_APP_PRODUCTION_TYPE === 'WGT')
    if (needCopy) {
      // 复制非kt资源
      fs.copySync(cachePluginDir, join(outputDir, pluginRelativeDir), {
        filter(src) {
          return !src.endsWith('app-android')
        },
      })
      // copy kt to src
      fs.copySync(
        path.resolve(cachePluginDir, 'utssdk', 'app-android'),
        join(outputDir, pluginRelativeDir, 'utssdk', 'app-android', 'src'),
        {
          filter(src) {
            if (fs.statSync(src).isDirectory()) return true
            return src.endsWith('.kt')
          },
          overwrite: false,
        }
      )
      // 需要把 kt 文件放到 app-android/src 下
    }
    const artifacts = pkg.uni_modules?.artifacts
    const inject_apis = artifacts?.apis || []
    const scoped_slots = artifacts?.scoped_slots || []
    addPluginInjectApis(inject_apis)
    addPluginInjectComponents(artifacts?.components || [])
    return {
      dir: outputPluginDir,
      code: 'export default {}',
      deps: [] as string[],
      encrypt: true,
      inject_apis,
      scoped_slots,
    }
  }
  // development
  if (process.env.UNI_HBUILDERX_PLUGINS) {
    const { DUM } = requireUniHelpers()

    const ktFiles: Record<string, string> = sync('**/*.kt', {
      absolute: false,
      cwd: cachePluginDir,
    }).reduce((files, file) => {
      files[path.resolve(cachePluginDir, file)] = path.resolve(
        process.env.UNI_OUTPUT_DIR,
        pluginRelativeDir,
        file
      )
      return files
    }, {})
    const errMsg = await DUM(path.basename(pluginRelativeDir), ktFiles)
    if (errMsg) {
      console.error(errMsg)
      process.exit(0)
    }

    const jarFile = resolveJarPath(
      'app-android',
      cacheDir,
      pluginRelativeDir,
      'index.kt'
    )
    const waiting = { done: undefined }
    const kotlinFiles = Object.values(ktFiles)
    const uniModules =
      require(path.resolve(pluginDir, 'package.json')).uni_modules
        ?.dependencies || []

    const depJars = uniModules.length
      ? getUniModulesCacheJars(cacheDir, uniModules) // 依赖的jar
      : []

    const compilerServer = getCompilerServer<KotlinCompilerServer>(
      'uniapp-runextension'
    )
    if (!compilerServer) {
      throw new Error(`项目使用了uts插件，正在安装 uts Android 运行扩展...`)
    }
    const { code, msg } = await compileAndroidDex(
      true,
      compilerServer,
      kotlinFiles,
      jarFile,
      '',
      depJars,
      createStderrListener(outputDir, resolveSourceMapPath(), waiting, (m) => {
        if (m.file) {
          const file = normalizePath(m.file)
          if (file.startsWith(outputPluginDir)) {
            // 过滤部分插件的非错误信息
            if (m.type !== 'error') {
              return ''
            } else {
              // 移除完整路径
              m.file = pluginDir
            }
          }
        }
        return hbuilderFormatter(m)
      })
    )
    // 等待 stderrListener 执行完毕
    if (waiting.done) {
      await waiting.done
    }
    try {
      kotlinFiles.forEach((file) => {
        fs.removeSync(file)
      })
    } catch (e) {}
    if (!code) {
      const dexFile = resolveDexFile(jarFile)
      if (fs.existsSync(dexFile)) {
        restoreDex(pluginRelativeDir, cacheDir, outputDir, true)
      }
    } else if (msg) {
      console.error(msg)
    }

    const assets = path.resolve(cachePluginDir, 'assets')
    if (fs.existsSync(assets)) {
      fs.copySync(assets, path.resolve(outputDir, pluginRelativeDir, 'assets'))
    }
  }

  return {
    dir: outputPluginDir,
    code: 'export default {}',
    deps: [] as string[],
    encrypt: true,
    inject_apis: [],
    scoped_slots: [],
  }
}
