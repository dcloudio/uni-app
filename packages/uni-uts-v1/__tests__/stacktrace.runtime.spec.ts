import fs from 'fs-extra'
import path from 'path'

import { parseRuntimeStacktrace } from '../src/stacktrace'
import { normalizePath } from '../src/shared'

describe('uts:stacktrace:runtime', () => {
  // 添加 beforeAll 来设置 HTTP mock
  beforeAll(() => {
    global.fetch = jest.fn((url) =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve({}),
        text: () => {
          // mp-weixin
          if (url.includes('appservice/pages/index/index.js')) {
            return fs.readFileSync(
              path.resolve(__dirname, 'data/mp-weixin.js'),
              'utf8'
            )
            // mp-baidu
          } else if (url.includes('output/pages/index/index.js.map')) {
            return fs.readFileSync(
              path.resolve(__dirname, 'data/mp-baidu.js.map'),
              'utf8'
            )
            // mp-alipay
          } else if (url.includes('index.worker.js.map')) {
            const dir = normalizePath(
              path.resolve(
                __dirname,
                'examples/uni-app-x/output/dist/dev/.sourcemap/mp-alipay'
              )
            )
            return fs
              .readFileSync(
                path.resolve(__dirname, 'data/mp-alipay.js.map'),
                'utf8'
              )
              .replaceAll('{OUTPUT_ROOT}', dir)
            // mp-alipay
          } else if (url.includes('index.worker.js')) {
            return `//# sourceMappingURL=http://localhost:6600/sourcemaps/index.worker.js.map`
            // mp-toutiao
          } else if (url.includes('app-dist/pages/index/index.js')) {
            return fs.readFileSync(
              path.resolve(__dirname, 'data/mp-toutiao.js'),
              'utf8'
            )
          }
        },
      })
    ) as jest.Mock
  })

  // 每个测试后清理 mock
  afterEach(() => {
    jest.clearAllMocks()
  })

  // 所有测试完成后恢复原始 fetch
  afterAll(() => {
    jest.restoreAllMocks()
  })

  test('parseUTSKotlinRuntimeStacktrace', async () => {
    const cacheDir = path.resolve(
      __dirname,
      'examples/uni-app-x/output/cache/app-android'
    )
    expect(
      await parseRuntimeStacktrace(
        `java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
at java.util.ArrayList.get(ArrayList.java:437)
at uni.UNIXXXXXXX.GenPagesIndexIndex$1.invoke(index.kt:24)
at java.lang.reflect.Method.invoke(Native Method)
at io.dcloud.uniapp.vue.shared.IndexKt$callFunction$invoke$1.invoke(index.kt:627)
at io.dcloud.uniapp.vue.shared.IndexKt.callFunction(index.kt:638)`,
        {
          platform: 'app-android',
          language: 'kotlin',
          appid: '__UNI__XXXXXXX',
          cacheDir,
        }
      )
    ).toMatchSnapshot()
    expect(
      await parseRuntimeStacktrace(
        `java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
at java.util.ArrayList.get(ArrayList.java:437)
at uni.UNIXXXXXXX.GenPagesIndexIndex$click$1.invoke(index.kt:59)
at java.lang.reflect.Method.invoke(Native Method)
at io.dcloud.uniapp.vue.shared.IndexKt$callFunction$invoke$1.invoke(index.kt:627)
at io.dcloud.uniapp.vue.shared.IndexKt.callFunction(index.kt:638)`,
        {
          platform: 'app-android',
          language: 'kotlin',
          appid: '__UNI__XXXXXXX',
          cacheDir,
        }
      )
    ).toMatchSnapshot()
    expect(
      await parseRuntimeStacktrace(
        `java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
at java.util.ArrayList.get(ArrayList.java:437)
at uni.UNIXXXXXXX.IndexKt.test(index.kt:40)
at uni.UNIXXXXXXX.GenPagesIndexIndex$click1$1.invoke(index.kt:63)
at java.lang.reflect.Method.invoke(Native Method)
at io.dcloud.uniapp.vue.shared.IndexKt$callFunction$invoke$1.invoke(index.kt:627)
at io.dcloud.uniapp.vue.shared.IndexKt.callFunction(index.kt:638)`,
        {
          platform: 'app-android',
          language: 'kotlin',
          appid: '__UNI__XXXXXXX',
          cacheDir,
        }
      )
    ).toMatchSnapshot()
    expect(
      await parseRuntimeStacktrace(
        `java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
at java.util.ArrayList.get(ArrayList.java:437)
at uni.UNIXXXXXXX.IndexKt.test(index.kt:40)`,
        {
          platform: 'app-android',
          language: 'kotlin',
          appid: '__UNI__XXXXXXX',
          cacheDir,
          logType: 'warn',
        }
      )
    ).toMatchSnapshot()
    expect(
      await parseRuntimeStacktrace(
        `java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
at java.util.ArrayList.get(ArrayList.java:437)
at uni.UNIXXXXXXX.IndexKt.test(index.kt:40)`,
        {
          platform: 'app-android',
          language: 'kotlin',
          appid: '__UNI__XXXXXXX',
          cacheDir,
          logType: 'error',
        }
      )
    ).toMatchSnapshot()

    expect(
      await parseRuntimeStacktrace(
        `java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Boolean
at uni.UNIXXXXXXX.GenComponentsTestBooleanTestBoolean.getDisabled1(test-boolean.kt:68)
at uni.UNIXXXXXXX.GenComponentsTestBooleanTestBoolean$1.invoke(test-boolean.kt:21)
at java.lang.reflect.Method.invoke(Native Method)
at io.dcloud.uniapp.vue.shared.IndexKt$callFunction$invoke$1.invoke(index.kt:708)`,
        {
          platform: 'app-android',
          language: 'kotlin',
          appid: '__UNI__XXXXXXX',
          cacheDir,
          logType: 'error',
        }
      )
    ).toMatchSnapshot()
  })
  test('parseUTSJavaScriptRuntimeStacktrace', async () => {
    const cacheDir = path.resolve(
      __dirname,
      'examples/uni-app-x/output/cache/app-ios'
    )
    expect(
      await parseRuntimeStacktrace(
        `app-service.js(5:60) ReferenceError:Can't find variable: a @app-service.js:5:60`,
        {
          platform: 'app-ios',
          language: 'javascript',
          cacheDir,
        }
      )
    ).toMatchSnapshot()
    expect(
      await parseRuntimeStacktrace(
        `at <Index __pageId=1 __pagePath="pages/index/index" __pageQuery=  ... >
Can't find variable: a
onLoad@app-service.js:9:64
callWithErrorHandling@uni-app-x-framework.js:2279:23
callWithAsyncErrorHandling@uni-app-x-framework.js:2286:38
@uni-app-x-framework.js:4763:45`,
        {
          platform: 'app-ios',
          language: 'javascript',
          cacheDir,
        }
      )
    ).toMatchSnapshot()
  })

  test('parseWeixinRuntimeStacktrace', async () => {
    const outputDir = path.resolve(
      __dirname,
      'examples/uni-app-x/output/dist/dev/mp-weixin'
    )
    expect(
      await parseRuntimeStacktrace(
        `Error: bbbbbbbbbbb12
at Proxy.throwError (http://127.0.0.1:36394/appservice/pages/index/index.js:10:13)`,
        {
          outputDir,
          platform: 'mp-weixin',
          language: 'javascript',
          cacheDir: '',
        }
      )
    ).toMatchSnapshot()
  })

  test('parseBaiduRuntimeStacktrace', async () => {
    const outputDir = path.resolve(
      __dirname,
      'examples/uni-app-x/output/dist/dev/mp-baidu'
    )
    expect(
      await parseRuntimeStacktrace(
        `Error: bbbbbbbbbbb12
at Proxy.throwError (http://127.0.0.1:36394/output/pages/index/index.js:11:13)`,
        {
          outputDir,
          platform: 'mp-baidu',
          language: 'javascript',
          cacheDir: '',
        }
      )
    ).toMatchSnapshot()
  })

  test('parseAlipayRuntimeStacktrace', async () => {
    const outputDir = path.resolve(
      __dirname,
      'examples/uni-app-x/output/dist/dev/mp-alipay'
    )
    expect(
      await parseRuntimeStacktrace(
        `Error: bbbbbbbbbbb12
at Proxy.throwError  (http://127.0.0.1:54378/index.worker.js?hash=66189eae:196:13)`,
        {
          outputDir,
          platform: 'mp-alipay',
          language: 'javascript',
          cacheDir: '',
        }
      )
    ).toMatchSnapshot()
  })

  test('parseToutiaoRuntimeStacktrace', async () => {
    const outputDir = path.resolve(
      __dirname,
      'examples/uni-app-x/output/dist/dev/mp-toutiao'
    )
    expect(
      await parseRuntimeStacktrace(
        `Error: bbbbbbbbbbb12
at Proxy.throwError (pages/index/index.js:26:19)
at pages/index/index.js:32:71
at callWithErrorHandling (common/vendor.js:1561:23)
at callWithAsyncErrorHandling (common/vendor.js:1568:19)
at invoke (common/vendor.js:4589:26)
at http://127.0.0.1:7046/core/3.46.0.17/tma-core.js:2:191115
at Ct (http://127.0.0.1:7046/core/3.46.0.17/tma-core.js:2:9631)
at t (http://127.0.0.1:7046/core/3.46.0.17/tma-core.js:2:15565)`,
        {
          outputDir,
          platform: 'mp-toutiao',
          language: 'javascript',
          cacheDir: '',
        }
      )
    ).toMatchSnapshot()
  })
})
