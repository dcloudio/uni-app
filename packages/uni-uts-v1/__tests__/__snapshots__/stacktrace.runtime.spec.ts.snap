// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uts:stacktrace:runtime parseAlipayRuntimeStacktrace 1`] = `
"Error: bbbbbbbbbbb12
at pages/index/index.vue:21:16
19 |                  ]))
20 |  
21 |                  throw new Error('bbbbbbbbbbb12')
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
22 |              }
23 |          }"
`;

exports[`uts:stacktrace:runtime parseBaiduRuntimeStacktrace 1`] = `
"Error: bbbbbbbbbbb12
at pages/index/index.vue:21:16
19 |                  ]))
20 |  
21 |                  throw new Error('bbbbbbbbbbb12')
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
22 |              }
23 |          }"
`;

exports[`uts:stacktrace:runtime parseToutiaoRuntimeStacktrace 1`] = `
"Error: bbbbbbbbbbb12
at pages/index/index.vue:21:16
19 |                  ]))
20 |  
21 |                  throw new Error('bbbbbbbbbbb12')
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
22 |              }
23 |          }"
`;

exports[`uts:stacktrace:runtime parseUTSJavaScriptRuntimeStacktrace 1`] = `
"error: [EXCEPTION] ReferenceError:Can't find variable: a
at pages/index/index.uvue:3:8
1  |  <script lang="uts">
2  |      setTimeout(() => {
3  |          uni.__log__('log','at pages/index/index.uvue:3',a)
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
4  |      }, 1000)
5  |      export default {"
`;

exports[`uts:stacktrace:runtime parseUTSJavaScriptRuntimeStacktrace 2`] = `
"error: [EXCEPTION] at <Index __pageId=1 __pagePath="pages/index/index" __pageQuery=  ... >
Can't find variable: a
at pages/index/index.uvue:7:12
5  |      export default {
6  |          onLoad() {
7  |              uni.__log__('log','at pages/index/index.uvue:7',a)
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
8  |          }
9  |      }"
`;

exports[`uts:stacktrace:runtime parseUTSKotlinRuntimeStacktrace 1`] = `
"error: java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
at pages/index/index.uvue:13:12
11 |          onLoad() {
12 |              const arr = [1]
13 |              console.log(arr[1])
   |              ^^^^^^^^^^^^^^^^^^^
14 |          },
15 |          methods: {"
`;

exports[`uts:stacktrace:runtime parseUTSKotlinRuntimeStacktrace 2`] = `
"error: java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
at pages/index/index.uvue:18:16
16 |              click() {
17 |                  const arr = [1]
18 |                  console.log(arr[1])
   |                  ^^^^^^^^^^^^^^^^^^^
19 |              },
20 |              click1() {"
`;

exports[`uts:stacktrace:runtime parseUTSKotlinRuntimeStacktrace 3`] = `
"error: java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
at pages/index/test.uts:3:4
1  |  export function test() {
2  |      const arr = [1]
3  |      console.log(arr[1])
   |      ^^^^^^^^^^^^^^^^^^^
4  |  }"
`;

exports[`uts:stacktrace:runtime parseUTSKotlinRuntimeStacktrace 4`] = `
"​error: java.lang.IndexOutOfBoundsException: Index: 1, Size: 1​
at pages/index/test.uts:3:4
1  |  export function test() {
2  |      const arr = [1]
3  |      console.log(arr[1])
   |      ^^^^^^^^^^^^^^^^^^^
4  |  }"
`;

exports[`uts:stacktrace:runtime parseUTSKotlinRuntimeStacktrace 5`] = `
"‌error: java.lang.IndexOutOfBoundsException: Index: 1, Size: 1‌
at pages/index/test.uts:3:4
1  |  export function test() {
2  |      const arr = [1]
3  |      console.log(arr[1])
   |      ^^^^^^^^^^^^^^^^^^^
4  |  }"
`;

exports[`uts:stacktrace:runtime parseUTSKotlinRuntimeStacktrace 6`] = `
"‌error: java.lang.ClassCastException: java.lang.String cannot be cast to java.lang.Boolean‌
at components/test-boolean/test-boolean.vue:19:12
17 |          },
18 |          created(){
19 |              console.log(this.disabled1)
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
20 |          }
21 |      }"
`;

exports[`uts:stacktrace:runtime parseWeixinRuntimeStacktrace 1`] = `
"Error: bbbbbbbbbbb12
at pages/index/index.vue:21:16
19 |                  ]))
20 |  
21 |                  throw new Error('bbbbbbbbbbb12')
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
22 |              }
23 |          }"
`;
